File: Max Area of Island - Leetcode 695/Max Area of Island - Leetcode 695.cpp
class Solution {
public:
    int maxAreaOfIsland(vector<vector<int>>& grid) {
        int m = grid.size();
        int n = grid[0].size();
        int maxArea = 0;

        for (int i = 0; i < m; ++i) {
            for (int j = 0; j < n; ++j) {
                if (grid[i][j] == 1) {
                    maxArea = max(maxArea, dfs(grid, i, j));
                }
            }
        }

        return maxArea;
    }

private:
    int dfs(vector<vector<int>>& grid, int i, int j) {
        // Directions for moving in the grid (right, left, down, up)
        static const vector<pair<int, int>> directions = {
            {0, 1}, {0, -1}, {1, 0}, {-1, 0}
        };

        if (i < 0 || i >= grid.size() || j < 0 || j >= grid[0].size() || grid[i][j] != 1) {
            return 0;
        }

        grid[i][j] = 0; // Mark as visited
        int area = 1;
        for (const auto& dir : directions) {
            int ni = i + dir.first;
            int nj = j + dir.second;
            area += dfs(grid, ni, nj);
        }
        return area;
    }
};

File: Max Area of Island - Leetcode 695/Max Area of Island - Leetcode 695.js
var maxAreaOfIsland = function(grid) {
    const m = grid.length;
    const n = grid[0].length;
    let maxArea = 0;

    const directions = [0, 1, 0, -1, 0]; // directions for moving right, down, left, up

    function dfs(i, j) {
        if (i < 0 || i >= m || j < 0 || j >= n || grid[i][j] !== 1) {
            return 0;
        }
        grid[i][j] = 0; // mark as visited
        let area = 1;
        for (let d = 0; d < 4; d++) {
            const ni = i + directions[d];
            const nj = j + directions[d + 1];
            area += dfs(ni, nj);
        }
        return area;
    }

    for (let i = 0; i < m; i++) {
        for (let j = 0; j < n; j++) {
            if (grid[i][j] === 1) {
                maxArea = Math.max(maxArea, dfs(i, j));
            }
        }
    }

    return maxArea;
};
